File: backend/app/routes/delivery_requests.py
from flask import Blueprint, jsonify, request
from flask_jwt_extended import jwt_required

from app.services.delivery_request_service import (
    cancel_delivery_request,
    create_delivery_request,
    get_delivery_request,
    list_customer_delivery_requests,
)
from app.utils.auth_decorators import resolve_jwt_user

delivery_requests_bp = Blueprint("delivery_requests", __name__)


@delivery_requests_bp.route("/me/delivery-requests", methods=["GET"])
@jwt_required()
def list_my_delivery_requests():
    user, auth_err = resolve_jwt_user()
    if auth_err:
        return auth_err

    requests = list_customer_delivery_requests(user)
    return jsonify({"delivery_requests": [item.to_dict(include_packages=True) for item in requests]})


@delivery_requests_bp.route("/me/delivery-requests", methods=["POST"])
@jwt_required()
def create_my_delivery_request():
    user, auth_err = resolve_jwt_user()
    if auth_err:
        return auth_err

    data = request.get_json(silent=True) or {}
    package_ids = data.get("package_ids") or []
    if not isinstance(package_ids, list):
        return jsonify({"error": "package_ids must be an array"}), 400

    delivery_address_id = data.get("delivery_address_id")
    if not delivery_address_id:
        return jsonify({"error": "delivery_address_id is required"}), 400

    try:
        delivery_request = create_delivery_request(
            user,
            package_ids=package_ids,
            delivery_address_id=delivery_address_id,
            notes=data.get("notes"),
        )
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({"delivery_request": delivery_request.to_dict(include_packages=True)}), 201


@delivery_requests_bp.route("/me/delivery-requests/<request_id>", methods=["DELETE"])
@jwt_required()
def cancel_my_delivery_request(request_id: str):
    user, auth_err = resolve_jwt_user()
    if auth_err:
        return auth_err

    delivery_request = get_delivery_request(request_id)
    if not delivery_request or delivery_request.customer_id != user.id:
        return jsonify({"error": "Delivery request not found"}), 404

    try:
        delivery_request = cancel_delivery_request(delivery_request, by_customer=True)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({"delivery_request": delivery_request.to_dict(include_packages=True)})


@delivery_requests_bp.route("/me/payment-total", methods=["POST"])
@jwt_required()
def compute_my_payment_total():
    user, auth_err = resolve_jwt_user()
    if auth_err:
        return auth_err

    data = request.get_json(silent=True) or {}
    package_ids = data.get("package_ids") or []
    if not isinstance(package_ids, list):
        return jsonify({"error": "package_ids must be an array"}), 400

    from app.services.delivery_request_service import compute_payment_total_with_delivery

    try:
        totals = compute_payment_total_with_delivery(user, package_ids)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify(totals)

File: server/src/routes/deliveryRequests.ts
import { Router } from 'express'
import { jwtRequired, requireAuth, type AuthRequest } from '../middleware/auth.js'
import { routeParam } from '../lib/routeParams.js'
import {
  cancelDeliveryRequest,
  computePaymentTotalWithDelivery,
  createDeliveryRequest,
  getDeliveryRequest,
  listCustomerDeliveryRequests,
} from '../services/deliveryRequestService.js'

export const deliveryRequestsRouter = Router()

deliveryRequestsRouter.get('/me/delivery-requests', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const requests = await listCustomerDeliveryRequests(req.user!)
  return res.json({ delivery_requests: requests })
})

deliveryRequestsRouter.post('/me/delivery-requests', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const data = req.body ?? {}
  const packageIds = data.package_ids ?? []
  if (!Array.isArray(packageIds)) return res.status(400).json({ error: 'package_ids must be an array' })
  if (!data.delivery_address_id) return res.status(400).json({ error: 'delivery_address_id is required' })

  try {
    const deliveryRequest = await createDeliveryRequest(req.user!, {
      packageIds,
      deliveryAddressId: String(data.delivery_address_id),
      notes: data.notes as string | undefined,
    })
    return res.status(201).json({ delivery_request: deliveryRequest })
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})

deliveryRequestsRouter.delete('/me/delivery-requests/:requestId', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const deliveryRequest = await getDeliveryRequest(routeParam(req.params.requestId))
  if (!deliveryRequest || deliveryRequest.customerId !== req.user!.id) {
    return res.status(404).json({ error: 'Delivery request not found' })
  }

  try {
    const updated = await cancelDeliveryRequest(deliveryRequest, true)
    return res.json({ delivery_request: updated })
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})

deliveryRequestsRouter.post('/me/payment-total', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  const packageIds = req.body?.package_ids ?? []
  if (!Array.isArray(packageIds)) return res.status(400).json({ error: 'package_ids must be an array' })

  try {
    const totals = await computePaymentTotalWithDelivery(req.user!, packageIds)
    return res.json(totals)
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})
